src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,548 @@
+            }
+        }
+    }
+
+    fn release_fault(queue: *GridRepairQueue, fault_index: usize) void {
+        assert(!queue.canceling);
+        assert(queue.faulty_blocks_repair_index < queue.faulty_blocks.count());
+
+        switch (queue.faulty_blocks.entries.items(.value)[fault_index].progress) {
+            .block => queue.enqueued_single_blocks -= 1,
+            .table_index => queue.enqueued_table_blocks -= 1,
+            .table_content => queue.enqueued_table_blocks -= 1,
+        }
+
+        queue.faulty_blocks.swapRemoveAt(fault_index);
+
+        if (queue.faulty_blocks_repair_index == queue.faulty_blocks.count()) {
+            queue.faulty_blocks_repair_index = 0;
+        } else {
+            if (queue.faulty_blocks_repair_index > fault_index) {
+                queue.faulty_blocks_repair_index -= 1;
+            }
+        }
+    }
+
+    pub fn cancel(queue: *GridRepairQueue) void {
+        assert(!queue.canceling);
+        assert(queue.checkpointing == null);
+
+        queue.faulty_blocks.clearRetainingCapacity();
+        queue.faulty_blocks_repair_index = 0;
+        queue.enqueued_single_blocks = 0;
+        queue.enqueued_table_blocks = 0;
+
+        queue.canceling = true;
+        while (queue.faulty_tables.pop()) |table| {
+            (table.callback)(table, .canceled);
+        }
+        queue.canceling = false;
+    }
+
+    pub fn checkpoint_commence(
+        queue: *GridRepairQueue,
+        free_set: *const vsr.superblock.SuperBlockFreeSet,
+    ) void {
+        assert(!queue.canceling);
+        assert(queue.checkpointing == null);
+        assert(queue.faulty_blocks.count() ==
+            queue.enqueued_single_blocks + queue.enqueued_table_blocks);
+
+        var aborting: usize = 0;
+
+        var faulty_blocks = queue.faulty_blocks.iterator();
+        while (faulty_blocks.next()) |fault_entry| {
+            const fault_address = fault_entry.key_ptr.*;
+            assert(!free_set.is_free(fault_address));
+            assert(fault_entry.value_ptr.state != .aborting);
+
+            if (free_set.is_released(fault_address)) {
+                switch (fault_entry.value_ptr.state) {
+                    .waiting => {
+                        faulty_blocks.index -= 1;
+                        faulty_blocks.len -= 1;
+                        queue.release_fault(faulty_blocks.index);
+                    },
+                    .writing => {
+                        fault_entry.value_ptr.state = .aborting;
+                        aborting += 1;
+                    },
+                    .aborting => unreachable,
+                }
+            }
+        }
+
+        var tables: FIFO(RepairTable) = .{ .name = queue.faulty_tables.name };
+        while (queue.faulty_tables.pop()) |table| {
+            assert(!free_set.is_free(table.index_address));
+
+            if (free_set.is_released(table.index_address)) {
+                (table.callback)(table, .released);
```

Comment:
Hm, this callback also feels complicated. Am i correct that this callback could cause the replica to call `enqueued_table`, so `enqueued_table` might be running while we are in hte middle of `checkpoint_commence`? Is there anyway  we could simplify that? Such re-entrancy I think goes against

> Use **only very simple, explicit control flow** for clarity.

:thinking: 

What if we remove table-synced callback, and instead ask the replica to explicitly poll for synced tables?

Reply:
Yeah, I guess we could poll in `grid_repair_block_callback` ?

Reply:
:thinking: So then `Replica` would not own the `RepairTable`s, `GridRepairQueue` would own them, and `Replica` would poll for their availability?

Fwiw the analogue right now is quite similar to e.g. `read_block`, where it is safe for the caller to call `read_block` within the callback of another `read_block`, reusing the same `*Read`.

Reply:
No, what I had in mind is the same ownership structure as today. Rather:

* we remove callback from repair table
* GridRepaiRQueue maintains a FIFO of repaired tables
* when the table is repaired, we push it onto this ready queue, instead of invoking calllback
* We add API to drain the ready tables


So the replica doesn’t get notified when a table is repaired. Rather, it reaps ready table at its own leisure.

Reply:
Ah, that makes sense

Reply:
Don’t react to stimuli basically — poll instead of callback. The question is, can the replica know when to poll?

Reply:
We could put it in the repair timeout.

Reply:
That, and I think callback from on_block would do the trick.

Reply:
>Fwiw the analogue right now is quite similar to e.g. read_block, where it is safe for the caller to call read_block within the callback of another read_block, reusing the same *Read.

I would worry about this as well. What makes me uneasy is that this is re-entrancy -- an object's method is called, while another method is executed, and this happens indirectly via callback. In theory, this is bad because it might cause unanticipated violation of the invariants: inside the first method, invariants might be _temporarily_ suspended, but we might expect the invariants to hold when we enter the nested-through-callback method call. In practice, I noticed that two-thirds of unsound `unsafe` blocks in Rust look like this:

```
fn allegedly_safe_abstraction(callback: fn) {
  unsafe {
   ...
   callback()
   ...
  }
}
```

where UB happens if the (safe) user-supplied callback manages to call `allegedly_safe_abstraction` re-entrantly.